be/src/exec/pipeline/scan/chunk_buffer_limiter.h

```diff
@@ -0,0 +1,126 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#pragma once
+
+#include <atomic>
+#include <memory>
+#include <mutex>
+
+namespace starrocks::pipeline {
+
+class ChunkBufferToken;
+using ChunkBufferTokenPtr = std::unique_ptr<ChunkBufferToken>;
+class ChunkBufferLimiter;
+using ChunkBufferLimiterPtr = std::unique_ptr<ChunkBufferLimiter>;
+
+class ChunkBufferToken {
+public:
+    virtual ~ChunkBufferToken() = default;
+};
+
+// Limit the capacity of a chunk buffer.
+// - Before creating a new chunk, should use `pin()` to pin a position in the buffer and return a token.
+// - After a chunk is popped from buffer, should desctruct the token to unpin the position.
+// All the methods are thread-safe.
+class ChunkBufferLimiter {
+public:
+    virtual ~ChunkBufferLimiter() = default;
+
+    // Update the chunk memory usage statistics.
+    // `added_sum_row_bytes` is the bytes of the new reading rows.
+    // `added_num_rows` is the number of the new read rows.
+    virtual void update_avg_row_bytes(size_t added_sum_row_bytes, size_t added_num_rows) = 0;
+
+    // Pin a position in the buffer and return a token.
+    // When desctructing the token, the position will be unpinned.
+    virtual ChunkBufferTokenPtr pin(int num_chunks) = 0;
+
+    // Returns true, when it cannot pin a position for now.
+    virtual bool is_full() const = 0;
+    // The number of already pinned positions.
+    virtual size_t size() const = 0;
+    // The max number of positions able to be pinned.
+    virtual size_t capacity() const = 0;
+    // The default capacity when there isn't chunk memory usage statistics.
+    virtual size_t default_capacity() const = 0;
+};
+
+// The capacity of this limiter is unlimited.
+class UnlimitedChunkBufferLimiter final : public ChunkBufferLimiter {
+public:
+    class Token final : public ChunkBufferToken {
+    public:
+        ~Token() override = default;
+    };
+
+public:
+    ~UnlimitedChunkBufferLimiter() override = default;
+
+    void update_avg_row_bytes(size_t added_sum_row_bytes, size_t added_num_rows) override {}
+
+    ChunkBufferTokenPtr pin(int num_chunks) override { return std::make_unique<Token>(); }
+
+    bool is_full() const override { return false; }
```

Comment:
Why the is_full is false, but size and capacity is 0?

Reply:
`capacity()` and `size()` are meaningless for `UnlimitedChunkBuffer`, but it is a little misunderstanding.
Therefore, I update it, make `size()` return the number of pinned tokens and `capacity` return `INT64_MAX`.